### ISSUE TITLE
Graceful Timeout Step Modifier

### ISSUE THREAD

**Issue by king-jam** [concourse/concourse#3737]
## What challenge are you facing?
When using the timeout step modifier and the time limit is reached, the task is interrupted/exited without any graceful termination option.

This is problematic when using an infrastructure provisioning tool like Terraform which does not write the state back unless it gracefully exits. Therefore, we cannot do an on_abort/on_error/on_failure task to cleanup the half created resources.

## What would make this better?
The ability to annotate the current `timeout` step modifier with a similar interface to:

http://man7.org/linux/man-pages/man1/timeout.1.html

My current workaround is to use the above and not use the timeout capability built into Concourse itself.

The proposal would be to enable the follow:
```
timeout: duration
```
OR
```
timeout: { 
         duration: duration,
         signal: signal,
         signal_timeout: duration
}
```
This is a first pass thought, alternatively, potentially allowing `timeout` to be overridden by a `step` could achieve a similar behavior?

## Are you interested in implementing this yourself?
Yes, I would assist with implementation if there are no major concerns with:
1. Potentially breaking backward compatibility (would try not to)
2. Providing this functionality.

**Comment by vito** [comment]
Makes sense! I like the pattern of allowing simple + complex `timeout:` configuration - it follows on the pattern we've set in place via #3580.

It might be best to avoid configuring signals and such as that brings up platform compatibility concerns (i.e. Darwin/Windows). Would something like this be sufficient?

```yaml
timeout:
  duration: 10s
  kill_after: 5m
```

On Linux this would mean 'send SIGTERM, wait 5 minutes, and then SIGKILL'. Implementation wise I think this would look something like this:

```go
process.Signal(garden.SignalTerminate)

select {
case <-process.Wait():
case <-time.After(timeoutConfig.KillAfter):
}

// might as well just make this unconditional so we ensure all processes are stopped (not just the main one)
container.Stop(true)
```

Implementation wise, I think this may be kind of challenging. Right now the timeout step just sets a timeout on the `context.Context`, and the actual interrupting happens in the [`task` step](https://github.com/concourse/concourse/blob/1ab4ef238fe352ae0e8e82ba22fd4ceacfc3fc32/atc/exec/task_step.go#L282-L285). You may have to sneak the additional configuration values into the `context.Context` so the task step can interrupt according to the config.

**Comment by vito** [comment]
Follow-up question: would you also want this kind of graceful stopping logic when the build is being aborted?

If so, then maybe this should just be a configuration on the `task` step, composable to `timeout` but also able to be configured on its own. Rough idea:

```yaml
task: foo
timeout: 10m
interrupt_timeout: 5m
```

**Comment by king-jam** [comment]
To your first post, yes, I think you captured my expectations
appropriately. I do think the signaling should be heavily defaulted for
compatibility. I'm up for trying to do that implementation based on your
tips and pointers.

To your follow up question, I think we would need to enable this on most
hooks. A graceful abort is definitely desirable in some situations. We may
want to evaluate the concept of double abort behavior. Someone like, if you
abort an already aborting task, it interrupts immediately.

**Comment by vito** [comment]
> We may want to evaluate the concept of double abort behavior. Something like, if you abort an already aborting task, it interrupts immediately.

Yeah, this would be nice. I think this would involve a lot more work, though. 🤔 Right now aborting is one-shot and acts like a state transition on the build (started -> 'aborting' flag -> aborted).

Either variation is fine with me. If we start with making this timeout-specific we can always generalize it to other steps later. 👍

**Comment by tjhiggins** [comment]
@king-jam could you reopen this issue? Would be a great feature for terraform tasks. Aborting/timeouts at the moment don't allow terraform to gracefully shutdown and maintains a state lock.

**Comment by king-jam** [comment]
@tjhiggins didn't realize this was needed by anyone else. I got super busy with work so didn't revisit.

My initial exploration found that this isn't as simple as it seems (unless you know the codebase well). The initial handling can go in `atc/worker/client.go` but I'm worried about unexpected cases of other context conditions signalling the same logic.

**Comment by tjhiggins** [comment]
The graceful aborts would be more important than timeouts. Should I create a separate feature request for that?

**Comment by king-jam** [comment]
That would fall under the same pieces of code. It would end up as a check
in whether it is a deadline or cancellation error that triggered it.

We can probably capture it under this and change the name to include both?

**Comment by tjhiggins** [comment]
Sounds great. Happy to help test whenever its prioritized.

**Comment by king-jam** [comment]
I submitted a PR.

**Comment by clarafu** [comment]
I think if we want to add an `interrupt_timeout` that also applies to build being aborted, we would want to let the user know when their aborted build is waiting on the `interrupt_timeout`. This would avoid the case when a user aborts a build that is configured with an `interrupt_timeout` of 10 minutes, currently they will see their build still in running state for 10 minutes before it shows that the build is aborted.

I'm not too sure what it should look like? Possibly some sort of processing/waiting state on the build page after they clicked the abort button?

**Comment by king-jam** [comment]
I agree. I didn't do nearly enough with the UI in this so there is some
curling around that I need to do. If imagine a spinner and a countdown but
I think that would require a fair bit more work because default context
timeouts couldn't be used.

**Comment by vito** [comment]
Maybe we should surface the `aborted` column from the database to the API and allow the web UI to indicate that a build is in that "aborting" state. That column gets set to `true` immediately and kicks off the aborting process, but I don't think we expose it at all:

https://github.com/concourse/concourse/blob/bc2ff0c9a606814a0a5d6338f44d5cd942b74456/atc/db/build.go#L581-L592

We'd probably also need some way for the web UI to notice this state change immediately, since when you're on the build page it'll be relying on the event stream for updates. Maybe we could emit an event for this? "abort started"? 🤔

**Comment by king-jam** [comment]
@vito I'll work with @clarafu and @xtreme-sameer-vohra to get this included. I suspect this will be a more involved change than I originally expected.

**Comment by tjhiggins** [comment]
@king-jam any update here? still a big pain point for us in regards to terraform jobs.

**Comment by king-jam** [comment]
I've been swamped with my other jobs unfortunately. More than happy to do
handoff here.
